lib/multicell_location/services/skyhook_integration.c

```diff
@@ -0,0 +1,329 @@
+			 cell_data->current_cell.mnc,
+			 cell_data->current_cell.tac,
+			 cell_data->current_cell.id,
+			 cell_data->current_cell.phys_cell_id,
+			 cell_data->current_cell.timing_advance,
+			 get_rsrp(cell_data->current_cell.rsrp),
+			 cell_data->current_cell.earfcn);
+		if ((len < 0) || (len >= sizeof(body))) {
+			LOG_ERR("Too small buffer for HTTP request body");
+			return -ENOMEM;
+		}
+
+		len = snprintk(buf, buf_len, HTTP_REQUEST_HEADER "%s", imei, strlen(body), body);
+		if ((len < 0) || (len >= buf_len)) {
+			LOG_ERR("Too small buffer for HTTP request body");
+			return -ENOMEM;
+		}
+
+		return 0;
+	}
+
+	*neighbors = 0;
+
+	for (size_t i = 0; i < cell_data->ncells_count; i++) {
+		char element[100];
+		len = snprintk(element, sizeof(element), HTTP_REQUEST_BODY_NEIGHBOR_ELEMENT "%s",
+			 mode == LTE_LC_LTE_MODE_LTEM ? "lte" : "nbiot",
+			 cell_data->neighbor_cells[i].phys_cell_id,
+			 get_rsrp(cell_data->current_cell.rsrp),
+			 cell_data->neighbor_cells[i].earfcn,
+			 i + 1 < cell_data->ncells_count ? "," : "");
+		if ((len < 0) || (len >= sizeof(element))) {
+			LOG_ERR("Too small buffer for cell element");
+			return -ENOMEM;
+		}
+
+		strcat(neighbors, element);
+	}
+
+	len = snprintk(body, sizeof(body), HTTP_REQUEST_BODY,
+		 mode == LTE_LC_LTE_MODE_LTEM ? "lte" : "nbiot",
+		 cell_data->current_cell.mcc,
+		 cell_data->current_cell.mnc,
+		 cell_data->current_cell.tac,
+		 cell_data->current_cell.id,
+		 cell_data->current_cell.phys_cell_id,
+		 cell_data->current_cell.timing_advance,
+		 get_rsrp(cell_data->current_cell.rsrp),
+		 cell_data->current_cell.earfcn,
+		 neighbors);
+	if ((len < 0) || (len >= sizeof(body))) {
+		LOG_ERR("Too small buffer for HTTP request body");
+		return -ENOMEM;
+	}
+
+	len = snprintk(buf, buf_len, HTTP_REQUEST_HEADER "%s", imei, strlen(body), body);
+	if ((len < 0) || (len >= buf_len)) {
+		LOG_ERR("Too small buffer for HTTP request");
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+int location_service_parse_response(const char *response, struct multicell_location *location)
+{
+	int err;
+	struct cJSON *root_obj, *location_obj, *lat_obj, *lng_obj, *accuracy_obj;
+	char *json_start, *http_status;
+
+	if ((response == NULL) || (location == NULL)) {
+		return -EINVAL;
+	}
+
+	/* The expected response format is the following:
+	 *
+	 * HTTP/1.1 <HTTP status, 200 OK if successful>
+	 * <Additional HTTP header elements>
+	 * <\r\n\r\n>
+	 * {"location":{"lat":<double>,"lng":<double>},"accuracy":<double>}
+	 */
+
+	http_status = strstr(response, "HTTP/1.1 200");
+	if (http_status == NULL) {
+		LOG_ERR("HTTP status was not 200");
+		return -1;
+	}
+
+	json_start = strstr(response, "\r\n\r\n");
+	if (json_start == NULL) {
+		LOG_ERR("No payload found");
+		return -1;
+	}
+
+	root_obj = cJSON_Parse(json_start);
+	if (root_obj == NULL) {
+		LOG_ERR("Could not parse JSON in payload");
+		return -1;
+	}
+
+	location_obj = cJSON_GetObjectItemCaseSensitive(root_obj, "location");
```

Comment:
Potential future improvement: `here_integration.c` and skyhook_integration.c` share common parsing functionality. This could be abstracted to some common parsing library between the two and possible future integration layers that share similar encoding. For easier test-ability and to save some lines.

Reply:
Yes, will probably move over to cJSON later, the current snprintk-based stuff isn't optimal